pkg/kafka/policy.go

```diff
@@ -281,9 +222,42 @@ func (req *RequestMessage) ruleMatches(rule api.PortRuleKafka) bool {
 // rules. The function will return true if the policy allows the message,
 // otherwise false is returned.
 func (req *RequestMessage) MatchesRule(rules []api.PortRuleKafka) bool {
+	reqTopicsMap := map[string]bool{}
+
+	// Maintain a map of all topics in the request
+	// Only if all the topics in a request are
+	// allowed by the rules, we should allow a request.
+	topics := req.GetTopics()
+	for _, topic := range topics {
+		reqTopicsMap[topic] = true
+	}
+
 	for _, rule := range rules {
-		if req.ruleMatches(rule) {
-			return true
+		switch len(topics) {
+		case 0:
+			if req.ruleMatches(rule, "") {
```

Comment:
I think this call with `matchTopic = ""` should only be made if `rule.Topic != ""`, to be consistent with the code below.